### ISSUE TITLE
sudo yarn global add on Debian (Raspbian) issue

### ISSUE THREAD

**Issue by jugglingcats** [yarnpkg/yarn#2807]
**Do you want to request a *feature* or report a *bug*?**

Bug. Apologies if this is a duplicate.

**What is the current behavior?**

```
$ sudo yarn global add nodemon
yarn global v0.21.3
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
warning fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Installed "nodemon@1.11.0" with binaries:
      - nodemon
warning No license field
Done in 10.63s.
```

```
$ nodemon
-bash: nodemon: command not found
```

Symbolic link is broken (as seen by non-root user):
```
$ file /usr/bin/nodemon
/usr/bin/nodemon: broken symbolic link to ../../root/.config/yarn/global/node_modules/nodemon/bin/nodemon.js
```

Because root folder is not readable by non-root user:
```
ls -ld /root
drwx------ 13 root root 4096 Feb 28 15:43 /root
```

**What is the expected behavior?**

After sudo install the command should be runnable as a normal user. I expect it's very normal on most linux distros for the root user's home directory to be non-readable except by root.

**Please mention your node.js, yarn and operating system version.**

```
$ node -v
v7.5.0

$ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 8 (jessie)"
NAME="Raspbian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

**Comment by jugglingcats** [comment]
Tried making /root readable but get further errors, eg:

```
$ nodemon
fs.js:558
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: EACCES: permission denied, open '/root/.config/yarn/global/node_modules/arr-flatten/index.js'
    at Object.fs.openSync (fs.js:558:18)
    at Object.fs.readFileSync (fs.js:468:33)
    at Object.Module._extensions..js (module.js:579:20)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/root/.config/yarn/global/node_modules/arr-diff/index.js:10:15)
    at Module._compile (module.js:571:32)
```

**Comment by Daniel15** [comment]
This is odd... If running as root, it should be using `/usr/local/share/yarn` rather than `~/.yarn`. See https://github.com/yarnpkg/yarn/blob/32b62607e58f52bdbcca3d5069c968d4c5712714/src/util/user-home-dir.js#L5-L7

Can you see whether ROOT_USER is true in that file (you can just hack a `console.log` call into it)?

**Comment by jugglingcats** [comment]
Very odd. I added some logging to the file you suggested and `constants.js` which is `require`d by it. There is something funny with the order of execution, like the `require` is returning before being fully executed. In the log statement printing value of `_require` I expected to see exported constants, not functions?!

*user-home-dir.js* snippet
```
console.log("Requiring constants.js...");

var _require = require('../constants');

console.log("Require complete: _require=", _require);

const ROOT_USER = _require.ROOT_USER;

console.log("ROOT USER (in user-home-dir.js): ", ROOT_USER);

const userHomeDir = process.platform === 'linux' && ROOT_USER ? path.resolve('/usr/local/share') : require('os').homedir();

exports.default = userHomeDir;
```

*constants.js* snippet
```
function getUid() {
  if (process.platform !== 'win32' && process.getuid) {
    return process.getuid();
  }
  return null;
}

const ROOT_USER = exports.ROOT_USER = isRootUser(getUid());
console.log("Exported ROOT_USER (in constants.js), value=", ROOT_USER);

function isRootUser(uid) {
  return uid === 0;
}
```

*output*
```
# yarn global add nodemon
Requiring constants.js...
Require complete: _require= { getPathKey: [Function: getPathKey],
  isRootUser: [Function: isRootUser] }
ROOT USER (in user-home-dir.js):  undefined
Exported ROOT_USER (in constants.js), value= true
yarn global v0.21.3
...
```

Note nodejs version v7.5.0.

**Comment by jugglingcats** [comment]
I reproduced this issue on PIXEL for PC (x86 live ISO distro of Raspbian) running nodejs v6.10.0. Same yarn version v0.21.3.

I downloaded and ran the PIXEL ISO in Virtual Box (http://downloads.raspberrypi.org/pixel_x86/images/pixel_x86-2016-12-13/2016-12-13-pixel-x86-jessie.iso). Then I followed instructions for Node v6 install (https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions) and Yarn package install (https://yarnpkg.com/lang/en/docs/install/#linux-tab).

Then I ran `yarn global add nodemon` and had the same broken symlink issue.

**Comment by Daniel15** [comment]
>  There is something funny with the order of execution

@connesc pointed out that there's a circular dependencies (as a comment on https://github.com/yarnpkg/yarn/pull/2635#pullrequestreview-25131333) which probably explains the odd behaviour. I think that needs to be fixed.

**Comment by connesc** [comment]
Sorry, I did not noticed this report. Indeed, I found the above bug while trying to understand this issue.
A possible fix is to move `ROOT_USER` to a separate module.

**Comment by bestander** [comment]
Should be fixed now

**Comment by connesc** [comment]
Will this be included in the next release? Any ETA ?

**Comment by bestander** [comment]
Yes, next release.
If you really want it sooner create a PR with a cherry pick onto 0.22-stable branch.
It'll come with the patch

**Comment by bestander** [comment]
Next release is in about 2 weeks

**Comment by connesc** [comment]
Thanks @bestander! I just created the PR on 0.22-stable.

**Comment by hbakhtiyor** [comment]
@bestander i've latest version (0.21.3) and have the same issue.

**Comment by Fohdeesha** [comment]
same issue on latest version as well  v0.22.0

**Comment by connesc** [comment]
@bestander: I guess [these commits](https://github.com/yarnpkg/yarn/compare/v0.22.0...0.22-stable) have been forgotten by mistake.

**Comment by connesc** [comment]
Fixed in v0.23, thx! :tada:
